File: src/main/java/com/leijendary/spring/microservicetemplate/error/ResourceNotFoundExceptionHandler.java
package com.leijendary.spring.microservicetemplate.error;

import com.leijendary.spring.microservicetemplate.data.response.ErrorResponse;
import com.leijendary.spring.microservicetemplate.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import static java.util.Locale.getDefault;
import static org.springframework.http.HttpStatus.NOT_FOUND;

@RestControllerAdvice
@Order(3)
@RequiredArgsConstructor
public class ResourceNotFoundExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(ResourceNotFoundException.class)
    @ResponseStatus(NOT_FOUND)
    public ErrorResponse catchResourceNotFound(final ResourceNotFoundException exception) {
        final var code = "error.resource.notFound";
        final var message = messageSource.getMessage(code,
                new Object[] { exception.getResource(), exception.getIdentifier() }, getDefault());

        return ErrorResponse.builder()
                .addError(exception.getResource(), code, message)
                .status(NOT_FOUND)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/spring/core/error/ResourceNotFoundExceptionHandler.kt
package com.leijendary.spring.core.error

import com.leijendary.spring.core.exception.ResourceNotFoundException
import com.leijendary.spring.core.model.ErrorModel
import com.leijendary.spring.core.util.RequestContext.locale
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
@Order(3)
class ResourceNotFoundExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(ResourceNotFoundException::class)
    @ResponseStatus(NOT_FOUND)
    fun catchResourceNotFound(exception: ResourceNotFoundException): List<ErrorModel> {
        val code = "error.resource.notFound"
        val message = messageSource.getMessage(code, emptyArray(), locale)
        val error = ErrorModel(id = exception.id.toString(), code = code, message = message, source = exception.source)

        return listOf(error)
    }
}
